lib/unison-sqlite/src/Unison/Sqlite/Transaction.hs

```diff
@@ -72,33 +81,110 @@ newtype Transaction a
+-- BEGIN/COMMIT statements.
+--
+-- The transaction is never retried, so it is (more) safe to interleave arbitrary IO actions.
+runWriteTransaction :: MonadUnliftIO m => Connection -> ((forall x. Transaction x -> m x) -> m a) -> m a
+runWriteTransaction conn f =
+  withRunInIO \runInIO ->
+    uninterruptibleMask \restore ->
+      runWriteTransaction_
+        restore
+        100_000
+        conn
+        (runInIO (f (\transaction -> liftIO (unsafeUnTransaction transaction conn))))
+{-# SPECIALIZE runWriteTransaction :: Connection -> ((forall x. Transaction x -> IO x) -> IO a) -> IO a #-}
+
+runWriteTransaction_ :: (forall x. IO x -> IO x) -> Int -> Connection -> IO a -> IO a
+runWriteTransaction_ restore microseconds conn transaction = do
+  keepTryingToBeginImmediate restore conn microseconds
+  result <- restore transaction `onException` ignoringExceptions (Connection.rollback conn)
+  Connection.commit conn
+  pure result
+
+-- @BEGIN IMMEDIATE@ until success.
+keepTryingToBeginImmediate :: (forall x. IO x -> IO x) -> Connection -> Int -> IO ()
+keepTryingToBeginImmediate restore conn =
+  let loop microseconds =
+        try @_ @SqliteQueryException (Connection.beginImmediate conn) >>= \case
+          Left SqliteBusyException -> do
+            restore (threadDelay microseconds)
+            loop (microseconds * 2)
+          Left exception -> throwIO exception
+          Right () -> pure ()
+   in loop
+
+ignoringExceptions :: IO () -> IO ()
+ignoringExceptions action =
+  action `catchAny` \_ -> pure ()
+
+-- | Unwrap the transaction newtype, throwing away the sending of BEGIN/COMMIT + automatic retry.
+unsafeUnTransaction :: Transaction a -> Connection -> IO a
+unsafeUnTransaction (Transaction action) =
+  action
+
+-- | Perform an atomic sub-computation within a transaction; if it returns 'Left', it's rolled back.
+savepoint :: Transaction (Either a a) -> Transaction a
+savepoint (Transaction action) = do
+  Transaction \conn -> do
+    -- Generate a random name for the savepoint, so the caller isn't burdened with coming up with a name. Seems
+    -- extremely unlikely for this to go wrong (i.e. some super nested withSavepoint call that ends up generating the
+    -- same savepoint name twice in a single scope).
+    name <- Text.pack <$> replicateM 10 (Random.randomRIO ('a', 'z'))
+    Connection.withSavepointIO conn name \rollback ->
+      action conn >>= \case
+        Left result -> do
+          rollback
+          pure result
         Right result -> pure result
-    Connection.execute_ conn "COMMIT"
-    pure result
-  where
-    rollback :: IO ()
-    rollback =
-      Connection.execute_ conn "ROLLBACK"
 
-    ignoringExceptions :: IO () -> IO ()
-    ignoringExceptions action =
-      action `catchAny` \_ -> pure ()
+-- | Perform IO inside a transaction, which should be idempotent, because it may be run more than once.
+-- FIXME rename to unsafeIO or something
+unsafeIO :: IO a -> Transaction a
+unsafeIO action =
```

Comment:
`unsafeIO` implies to me that it's something like `unsafePerformIO`; would something like `liftRepeatableIO` or `liftIdempotentIO` be clearer? It's not 'unsafe' to run IO in a transaction, it's just important that the IO be safely repeatable 😄